### PR TITLE
AYS-536 | User Update With Multiple Roles

### PR DESCRIPTION
**Pull request by htczrrsz**
## Pull Request

**Description:**
Added  updateUserWithMultipleRoles test to UserUpdateTest class and refactored some user role update tests

**Changes Made:**

1. Add: updateUserWithMultipleRoles method
Added a new test method, updateUserWithMultipleRoles to the UserUpdateTest class.
This new test checks the functionality of updating a user with multiple role IDs, improving the test coverage for scenarios where users have more than one role.

2. Refactor: updateUserWithInvalidRoleList and createUserWithInvalidRoleId tests
Updated the updateUserWithInvalidRoleList and createUserWithInvalidRoleId test methods to include null and empty list checks for roleIds.
The tests now handle cases where roleIds is null or an empty list, ensuring better validation and coverage for edge cases.

**Additional Information:**
Add any additional information that might be relevant to reviewers.

**Reviewer Instructions:**
Provide instructions or guidelines for reviewers to follow when assessing your pull request.

### Checklist

Before submitting your pull request, ensure the following:

- [x] **Title and Branch Naming Conventions:**
    - The pull request title follows the
      standard: [Pull Request Naming Conventions](https://github.com/afet-yonetim-sistemi/ays-be-api-automation/blob/main/CONTRIBUTING.md#pull-request-naming-conventions)
      .
    - The branch name follows one of the conventions outlined in
      the [Branch Naming Conventions](https://github.com/afet-yonetim-sistemi/ays-be-api-automation/blob/main/CONTRIBUTING.md#branch-naming-conventions)
      .

- [x] **Local Testing:**
    - I have added positive and negative test scenarios (if applicable)
    - All tests are passing.
    - I tagged all test as smoke or regression
    - I added newly added tests to related xml file in test suite

- [x] **Code Quality:**
    - The code is formatted according to the project's coding guidelines and style.
    - The code has been reviewed to ensure its quality.
    - The code does not contain any issues flagged by SonarLint.

- [x] **Documentation:**
    - Necessary documentation has been added or existing documentation has been updated.

- [x] **Reviewers and Assignees:**
    - Default reviewers have been assigned to this pull request.
    - Assignees have been added if necessary.

- [x] **Labels and Associations:**
    - No specific actions are required in the Labels and Associations section for this pull request.

- [x] **Reviewer Instructions:**
    - Provide instructions or guidelines for reviewers to follow when assessing your pull request.

### CODE REVIEW

File: src/test/java/org/ays/auth/datasource/UserRoleRelationDataSource.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+                return roleIds;
+            }
+
+        } catch (SQLException exception) {
+            throw new RuntimeException("Error fetching role IDs for user: " + userId, exception);
+        }
+    }
+
+}
```
Comment (agitrubard) [inline src/test/java/org/ays/auth/datasource/UserRoleRelationDataSource.java:40]: Bu metodu `UserDataSource`'a `findAllRoleIdsById(String id)` olarak taşıyabiliriz, ekstra bir data source oluşturmamız şart değil çünkü kullanıcının rollerini getiriyoruz

File: src/test/java/org/ays/auth/tests/UserCreateTest.java [hunk 2]
```diff
@@ -87,13 +87,13 @@ public void createUserWithInvalidEmailAddress(String emailAddress, AysErrorMessa
     }
 
     @Test(groups = {"Regression"}, dataProvider = "invalidRoleIdListData", dataProviderClass = AysDataProvider.class)
-    public void createUserWithInvalidRoleId(String id, AysErrorMessage errorMessage, String field, String type) {
+    public void createUserWithInvalidRoleId(List<String> roleIds, AysErrorMessage errorMessage, String field, String type) {
```
Comment (agitrubard) [inline src/test/java/org/ays/auth/tests/UserCreateTest.java:90]:
```suggestion
    public void createUserWithInvalidRoles(List<String> roleIds, AysErrorMessage errorMessage, String field, String type) {
```

File: src/test/java/org/ays/auth/tests/UserUpdateTest.java [hunk 3]
```diff
@@ -134,7 +175,7 @@ public void updateUserWithInvalidEmail(String emailAddress, AysErrorMessage erro
     }
 
     @Test(groups = {"Regression"}, dataProvider = "invalidRoleIdListData", dataProviderClass = AysDataProvider.class)
-    public void updateUserWithInvalidRoleList(String id, AysErrorMessage errorMessage, String field, String type) {
+    public void updateUserWithInvalidRoleList(List<String> roleIds, AysErrorMessage errorMessage, String field, String type) {
```
Comment (agitrubard) [inline src/test/java/org/ays/auth/tests/UserUpdateTest.java:178]:
```suggestion
    public void updateUserWithInvalidRoles(List<String> roleIds, AysErrorMessage errorMessage, String field, String type) {
```

**Review by agitrubard — CHANGES_REQUESTED** [review]

**Review by agitrubard — APPROVED** [review]

File: src/test/java/org/ays/auth/datasource/UserDataSource.java [hunk 4]
```diff
@@ -148,4 +150,28 @@ public static String findLastCreatedUserStatusByInstitutionId(String institution
...
+                }
+
+                return roleIds;
+            }
+
+        } catch (SQLException exception) {
+            throw new RuntimeException("Error fetching role IDs for user: " + id, exception);
+        }
+    }
+
```
Comment (cagla88) [inline src/test/java/org/ays/auth/datasource/UserDataSource.java:176]: Bu methodu mevcut `findLastCreatedUserIdByInstitutionId` methoduyla ilişkili olacak şekilde düzenlersek nasıl olur?

File: src/test/java/org/ays/auth/tests/UserUpdateTest.java [hunk 5]
```diff
@@ -45,6 +47,44 @@ public void updateUserSuccessfully() {
...
+        List<String> assignedRoleIds = UserDataSource.findAllRoleIdsById(userId);
+        List<String> expectedRoleIds = List.of(firstRoleId, secondRoleId);
+
+        response.then()
+                .spec(AysResponseSpecs.expectSuccessResponseSpec());
+
+        assertTrue(assignedRoleIds.containsAll(expectedRoleIds), "Assigned roles do not match expected roles.");
+        assertTrue(expectedRoleIds.containsAll(assignedRoleIds), "Unexpected roles are assigned to the user.");
+        assertEquals(assignedRoleIds.size(), expectedRoleIds.size(),
+                "Unexpected number of roles assigned to the user.");
```
Comment (cagla88) [inline src/test/java/org/ays/auth/tests/UserUpdateTest.java:85]: Burda response'lar alt alta assertionlarda yine List'lerden sonra gelse daha düzenli ve anlaşılır olabilir.
Birde equals ile kontrolü sadece rol sayısı için yapmışız sanırım, aynı olup olmamasıyla da ilgili bir assertion ekleyebiliriz.

**Review by cagla88 — CHANGES_REQUESTED** [review]

File: src/test/java/org/ays/auth/tests/UserUpdateTest.java (hunk 5 above)
Comment (htczrrsz) [reply src/test/java/org/ays/auth/tests/UserUpdateTest.java:85]: @cagla88 assertTrue containsAll'lar ayni olup olmadiklarini test ediyor aslinda, baska bir seyi mi kast ettin ?

File: src/test/java/org/ays/auth/datasource/UserDataSource.java (hunk 4 above)
Comment (htczrrsz) [reply src/test/java/org/ays/auth/datasource/UserDataSource.java:176]: @cagla88 o zaman son kullanicinin rollerini getirir sadece, bu sekilde olduğu zaman daha efektif kullanilabilir bence. Ama roleDataSource'a koymaliymisim bu metodu yanlis yerde olmuş.

File: src/test/java/org/ays/auth/tests/UserUpdateTest.java (hunk 5 above)
Comment (cagla88) [reply src/test/java/org/ays/auth/tests/UserUpdateTest.java:85]: true'lar ile aynı olup olmadıklarından ziyade içerip içermediğini kontrol ediyor aslında. Yani yapmak istediğini anladım hem içerdiğini hemde size ları aynı mı diye check ediyorsun ama bunun yerine sadece aynı olup olmadıklarını check etsek yeterli olur diye belirtmek istemiştim.

**Review by agitrubard — APPROVED** [review]

**Review by cagla88 — APPROVED** [review]
